actions/workflows/create_vm_role.yaml

```diff
@@ -1,100 +1,152 @@
-        hosts: "{{ec2_instance_private_ip}}"
-        cmd: "shutdown -r +1 &"
-      on-success: "wait_for_ssh_post_reboot"
-    -
-      name: "wait_for_ssh_post_reboot"
-      ref: "linux.wait_for_ssh"
-      params:
-        hostname: "{{ec2_instance_private_ip}}"
-        username: "stanley"
-        keyfile: "{{keyfile}}"
-        ssh_timeout: 30
-        retries: 10
-      on-success: "puppet_bootstrap"
-    -
-      name: "puppet_bootstrap"
-      ref: "st2cd.puppet_bootstrap"
-      params:
-        hosts: "{{ec2_instance_private_ip}}"
-        role: "{{role}}"
-  default: "get_subnet_id"
+version: '2.0'
+
+st2cd.create_vm_role:
+    description: Create VM role on AWS
+    type: direct
+    input:
+        - instance_type
+        - environment
+        - key_name
+        - keyfile
+        - hostname
+        - dns_zone
+        - role
+        - creator
+        - purpose
+        - distro
+    tasks:
+        get_subnet_id:
+            action: st2.kv.get
+            input:
+                key: <% $.environment %>_subnet
+            on-success:
+                - get_ami
+
+        get_ami:
+            action: st2.kv.get
+            input:
+                key: ami_<% $.distro %>
+            on-success:
+                - check_cname_if_it_already_exists
+
+        check_cname_if_it_already_exists:
+            action: aws.r53_zone_get_cname
+            input:
+                name: <% $.hostname %>.<% $.dns_zone %>
+                zone: <% $.dns_zone %>
+            on-success:
+                - run_instance: <% task(check_cname).result.result[0] = null %>
+                - fail: <% task(check_cname).result.result[0] != null %>
+
+        run_instance:
+            action: aws.ec2_run_instances
+            input:
+                image_id: <% task(get_ami).result.result %>
+                instance_type: <% $.instance_type %>
+                subnet_id: <% task(get_subnet_id).result.result %>
+                key_name: <% $.key_name %>
+            publish:
+                ec2_instance_id: <% task(run_instance).result.result[0].id %>
+                ec2_instance_private_ip: <% task(run_instance).result.result[0].private_ip_address %>
+                ec2_instance_private_dns: <% task(run_instance).result.result[0].private_dns_name %>
+            on-success:
+                - wait_for_instance
+
+        wait_for_instance:
+            action: aws.ec2_wait_for_state
+            input:
+                instance_id: <% $.ec2_instance_id %>
+                state: running
+            on-success:
+                - wait_for_ssh
+
+        wait_for_ssh:
+            action: linux.wait_for_ssh
+            input:
+                hostname: <% $.ec2_instance_private_ip %>
+                username: stanley
+                keyfile: <% $.keyfile %>
+                ssh_timeout: 20
+                retries: 30
+            on-success:
+                - add_tags
+            on-error:
+                - destroy_vm_on_failure
```

Comment:
> How come we don't want to destroy VM if wait_for_instance, add_tags and add_cname fail? Can you provide an explanation?

- `wait_for_instance`: If this fails that means the instance was not created  i.e no running state. what are we destroying then?

- `add_cname`: We have taken care of the duplicate use-case earlier in `check_cname_if_it_already_exists`. Any other reason for failure, if any, should be checked and `destroy_vm` will not help debug in this case.


-  `add_tags`: again this is `aws` task. If this fails. I would like to keep the state of failure intact and fix/mitigate error before destroying vm. 

I don't have any problem adding task-default for these 3. But it would be better to know the reason of failure than to hide it behind `destroy_vm`. 

> I think you still need to review how task-defaults work and what I mean to override for get_subnet_id, get_ami_id, and run_instance tasks. 

Actually, I still didn't get what you mean by adding task-defaults for these tasks. Do you want me to add on-error/on-complete for these tasks ?